feat: protect core with API keys

fontanierh opened the pull request:
## Description

https://github.com/dust-tt/dust/issues/6220

- This adds the ability to enforce passing a valid API key header (Authorization) for clients making calls to the `core` API.
- Valid API keys are defined as a k8s file secret mounted as a volume on `core` pods. As the file is JSON, it's is painful to edit through our usual simple key/value secret patching setup (as it requires escaping JSON)
- The file is a JSON array, where each element is an object with the keys `client_name` and `api_key`. This allows identifying clients from core, and also allows having (temporarily) several valid keys for a given client while we rotate keys.
- The secret file is loaded  from disk into a `OnceCell<Arc<HashMap<String, Vec<String>>>>`, which allows to easy lazy loading, immutability and thread-safe lock-free access.
- API Keys are validated via an axum middleware that can easily be plugged on our axum routers
- All of this is exposed as a module that can be used as-is in any of our rust binaries (also `oauth` and `sqlite-worker`)

## Risk

For now the functionality is disabled. The middleware is added in the `core` pipeline, but will not error for missing / invalid API keys. We will however log unauthenticated access, so we can properly migrate the clients before enabling on `core` side.
The file file secret has been created (with just one fake api key for now), and added to the deployment.

## Deploy Plan

- apply infra, then check that the file is indeed available on `core` pods
- deploy `core`

Will then start the work on migrating clients.

fontanierh commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ test.ts
 qdrant_storage
 .*.swp
 .rgignore
+core-api-keys.json
```
note: this will require having a file like this one somewhere, and having the path to it in you local core secrets (as `API_KEYS`)

spolu replied:
Not a big fan of that as it complexity the dev setup for interviews

We could fail open in the absence of file but that's not desirable from a security perspective...

Why not use a simple env var?

fontanierh replied:
Mostly because it's annoying to set using kube CLI due to escaping. 

I can also add an env var to disable check that we can set in dev, will be just another line in the env file, wdyt ?

spolu replied:
Yes that seems reasonable

flvndvd commented on core/src/api_keys.rs:
```diff
@@ -0,0 +1,80 @@
+
+    if let Some(auth_header) = req.headers().get("Authorization") {
+        let auth_header = auth_header.to_str().map_err(|_| StatusCode::UNAUTHORIZED)?;
+        if auth_header.starts_with("Bearer ") {
+            let provided_key = &auth_header[7..];
```
`nit` To avoid magic variable like `7`.

if let Some(provided_key) = auth_header.strip_prefix("Bearer ") {
    // TODO:
}

flvndvd reviewed (commented):
🔥 Do you have an example of what the file looks like? By "client," do you mean the entity consuming this service (e.g., front, front-edge, etc.), or is it something different? I'm not quite sure how the client name is being used.

Also, will we still need to use the "paintful patch" secret for this JSON, right?

fontanierh commented:
@flvndvd a file would looks something like:
```
[{ "client_name": "client1", "api_key": "Bearer key1" }]
```

> By "client," do you mean the entity consuming this service (e.g., front, front-edge, etc.)

yes I mean that

> Also, will we still need to use the "paintful patch" secret for this JSON, right?

You can re-create it from file. It's still not ideal (ideal would be to mount it from GCP secret manager I guess), but at least you don't have to think about escaping JSON in the CLI

spolu commented on core/src/api_keys.rs:
```diff
@@ -0,0 +1,80 @@
+            let provided_key = &auth_header[7..];
+            for (client_name, keys) in api_keys.iter() {
+                if keys.contains(&provided_key.to_string()) {
+                    info!("API key validated for client '{}'", client_name);
```
Let's not log for that, but add the identity to the req if possible?

spolu replied:
So that we can grab it in the tracing layer in `CoreRequestMakeSpan` ?

fontanierh replied:
Good point 👍

spolu commented on core/src/api_keys.rs:
```diff
@@ -0,0 +1,80 @@
+    if ENFORCE_API_KEY_REQUIRED {
+        Err(StatusCode::UNAUTHORIZED)
+    } else {
+        info!("API key not validated, but not required");
```
no log required

fontanierh replied:
For now it's useful, so I make sure no clients are making any unauthenticated requests anymore before I make it required (and remove this log)

fontanierh commented on core/src/api_keys.rs:
```diff
@@ -0,0 +1,80 @@
+static ENFORCE_API_KEY_REQUIRED: bool = false;
+
+type ApiKeyMap = Arc<HashMap<String, Vec<String>>>;
+static API_KEYS: OnceCell<ApiKeyMap> = OnceCell::const_new();
```
Well with the OnceCell it's one lock saved. IMHO it's better using OnceCell and it's a very common / standard rust lib so I don't really see a downside to having it. 
Do you feel strongly that we should move it to rw lock ?

fontanierh replied:
But actually, OnceCell is included in Tokio so no need for a dep

spolu replied:
kk

fontanierh commented on core/src/api_keys.rs:
```diff
@@ -0,0 +1,80 @@
+    api_key: String,
+}
+
+async fn init_api_keys() -> Result<ApiKeyMap> {
```
No because it is async. That's why a `OnceCell` is perfect here

spolu replied:
still it's a new import that could be avoided :/

spolu replied:
there is a block read_to_string in std which would be called only once here which is fine IMHO and would largely simplify the code?

fontanierh replied:
> largely simplify the code

It would save maximum 2 lines of code here

> still it's a new import that could be avoided :/

It would be the exact same number of imports because we'd need a lock instead

fontanierh replied:
And reading the file synchronously at a random time seems objectively worse than simply doing the right thing and using tokio no ?

spolu replied:
same imports # but one removed dependency. Everything would go in the lazy_static so you would not need your getter/setter and init function

spolu replied:
I agree that the sync read is not ideal though.

fontanierh replied:
> same imports # but one removed dependency.

Also same number of dependencie. OnceCell is coming from Tokio, which we already have

fontanierh commented:
@flvndvd  changed to `strip_prefix` 👍 

@spolu removed the dependency, added the disable env var, added the identity to the request and to the core span 👍
